## 1. Imports and loading data

In [21]:
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA as sklearn_pca

np.set_printoptions(formatter={'float':'{:0.4f}'.format})
pd.set_option('display.precision', 5)

filePath = "C:/Chang/NEU/courses/cs6220 datamining/project/dataset_diabetes/diabetic_data.csv"
# Read dataset using pandas
data = pd.read_csv(filePath, sep=',',header=0)

## 2. Preprocessing
This data set contains both numeric and nominal data types. For numeric data we need to do normalization, while for categorical data, we perform one-hot encoding.

In [29]:
# Drop encounter_id and patient_nbr. Drop weight and payer_code which has 97% missing data
df = data.drop(['encounter_id', 'patient_nbr', 'weight', 'payer_code'], axis = 1)

In [22]:
# TODO: Currently performed binary encoding. Should try data binning.
df['diag_1'] = np.where(df['diag_1'].str.contains('250'), 1, 0)
df['diag_2'] = np.where(df['diag_2'].str.contains('250'), 1, 0)
df['diag_3'] = np.where(df['diag_3'].str.contains('250'), 1, 0)

In [23]:
# Standardizing numeric data
from sklearn.preprocessing import StandardScaler
numData = df[['time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 
              'number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses']]
numData_std = StandardScaler().fit_transform(numData)

In [24]:
# Confirm that the numeric data are standardized
df_numData = pd.DataFrame(numData_std)
df_numData.describe()

0            1            2            3            4  \
count  1.01766e+05  1.01766e+05  1.01766e+05  1.01766e+05  1.01766e+05   
mean   2.35731e-14  3.24394e-16  4.23327e-15 -2.18372e-14  1.00082e-13   
std    1.00000e+00  1.00000e+00  1.00000e+00  1.00000e+00  1.00000e+00   
min   -1.13765e+00 -2.13963e+00 -7.85398e-01 -1.84827e+00 -2.91461e-01   
25%   -8.02651e-01 -6.14795e-01 -7.85398e-01 -7.40920e-01 -2.91461e-01   
50%   -1.32655e-01  4.59666e-02 -1.99162e-01 -1.25726e-01 -2.91461e-01   
75%    5.37341e-01  7.06728e-01  3.87074e-01  4.89467e-01 -2.91461e-01   
max    3.21732e+00  4.51881e+00  2.73202e+00  7.99483e+00  3.28509e+01   

                 5            6            7  
count  1.01766e+05  1.01766e+05  1.01766e+05  
mean   2.50927e-14 -4.31450e-14  1.28417e-13  
std    1.00000e+00  1.00000e+00  1.00000e+00  
min   -2.12620e-01 -5.03276e-01 -3.32160e+00  
25%   -2.12620e-01 -5.03276e-01 -7.35733e-01  
50%   -2.12620e-01 -5.03276e-01  2.98612e-01  
75%   -2.12620e-01  2.88579e-01  8.15784e-01  
max    8.14667e+01  1.61257e+01  4.43599e+00

In [25]:
df[['time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 
              'number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses']] = df_numData

In [26]:
# Apply one-hot encoding to categorical columns
oneHotData = pd.get_dummies(df, columns=['race', 'gender', 'age', 'admission_type_id', 'discharge_disposition_id',
                                         'admission_source_id', 'medical_specialty', 
                                         'diag_1', 
                                         'diag_2', 'diag_3', 'max_glu_serum', 'A1Cresult', 
                                         'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 
                                         'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 
                                         'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 
                                         'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin',
                                         'glipizide-metformin', 'glimepiride-pioglitazone', 
                                         'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 
                                         'diabetesMed', 'readmitted'],
                            prefix=['race', 'gender', 'age', 'admission_type_id', 'discharge_disposition_id', 
                                    'admission_source_id', 'medical_specialty', 
                                    'diag_1', 'diag_2', 'diag_3', 
                                    'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 
                                    'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 
                                    'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 
                                    'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin',
                                    'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 
                                    'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed',
                                    'readmitted'])
oneHotData.describe()

time_in_hospital  num_lab_procedures  num_procedures  num_medications  \
count       1.01766e+05         1.01766e+05     1.01766e+05      1.01766e+05   
mean        2.35731e-14         3.24394e-16     4.23327e-15     -2.18372e-14   
std         1.00000e+00         1.00000e+00     1.00000e+00      1.00000e+00   
min        -1.13765e+00        -2.13963e+00    -7.85398e-01     -1.84827e+00   
25%        -8.02651e-01        -6.14795e-01    -7.85398e-01     -7.40920e-01   
50%        -1.32655e-01         4.59666e-02    -1.99162e-01     -1.25726e-01   
75%         5.37341e-01         7.06728e-01     3.87074e-01      4.89467e-01   
max         3.21732e+00         4.51881e+00     2.73202e+00      7.99483e+00   

       number_outpatient  number_emergency  number_inpatient  \
count        1.01766e+05       1.01766e+05       1.01766e+05   
mean         1.00082e-13       2.50927e-14      -4.31450e-14   
std          1.00000e+00       1.00000e+00       1.00000e+00   
min         -2.91461e-01      -2.12620e-01      -5.03276e-01   
25%         -2.91461e-01      -2.12620e-01      -5.03276e-01   
50%         -2.91461e-01      -2.12620e-01      -5.03276e-01   
75%         -2.91461e-01      -2.12620e-01       2.88579e-01   
max          3.28509e+01       8.14667e+01       1.61257e+01   

       number_diagnoses        race_?  race_AfricanAmerican      ...        \
count       1.01766e+05  101766.00000          101766.00000      ...         
mean        1.28417e-13       0.02234               0.18877      ...         
std         1.00000e+00       0.14777               0.39132      ...         
min        -3.32160e+00       0.00000               0.00000      ...         
25%        -7.35733e-01       0.00000               0.00000      ...         
50%         2.98612e-01       0.00000               0.00000      ...         
75%         8.15784e-01       0.00000               0.00000      ...         
max         4.43599e+00       1.00000               1.00000      ...         

       metformin-rosiglitazone_Steady  metformin-pioglitazone_No  \
count                    101766.00000               101766.00000   
mean                          0.00002                    0.99999   
std                           0.00443                    0.00313   
min                           0.00000                    0.00000   
25%                           0.00000                    1.00000   
50%                           0.00000                    1.00000   
75%                           0.00000                    1.00000   
max                           1.00000                    1.00000   

       metformin-pioglitazone_Steady     change_Ch     change_No  \
count                    1.01766e+05  101766.00000  101766.00000   
mean                     9.82646e-06       0.46195       0.53805   
std                      3.13472e-03       0.49855       0.49855   
min                      0.00000e+00       0.00000       0.00000   
25%                      0.00000e+00       0.00000       0.00000   
50%                      0.00000e+00       0.00000       1.00000   
75%                      0.00000e+00       1.00000       1.00000   
max                      1.00000e+00       1.00000       1.00000   

       diabetesMed_No  diabetesMed_Yes  readmitted_<30  readmitted_>30  \
count    101766.00000     101766.00000    101766.00000    101766.00000   
mean          0.22997          0.77003         0.11160         0.34928   
std           0.42081          0.42081         0.31487         0.47675   
min           0.00000          0.00000         0.00000         0.00000   
25%           0.00000          1.00000         0.00000         0.00000   
50%           0.00000          1.00000         0.00000         0.00000   
75%           0.00000          1.00000         0.00000         1.00000   
max           1.00000          1.00000         1.00000         1.00000   

       readmitted_NO  
count   101766.00000  
mean         0.53912  
std          0.49847  
min          0.00000 

## 3. PCA

In [27]:
pca = sklearn_pca(n_components=10)
pca.fit_transform(oneHotData)

array([[-3.4099, -0.7666, -0.1605, ..., 0.7717, -0.4167, 0.7278],
       [0.2984, 0.3564, -1.2648, ..., -0.4434, -0.8187, 0.1261],
       [-0.9564, -0.3818, 1.9668, ..., -0.8102, -0.9108, -2.1515],
       ..., 
       [0.0449, 0.9767, -1.1813, ..., -1.1445, 0.6754, 0.4100],
       [2.0375, -0.2151, -0.1680, ..., 0.8678, -0.1103, -0.2349],
       [-0.9541, -0.5231, 0.2045, ..., 0.3305, 1.1552, -1.4892]])

In [28]:
print(pca.explained_variance_ratio_)

[0.1280 0.0868 0.0691 0.0597 0.0543 0.0506 0.0456 0.0408 0.0301 0.0288]


PC1 only explained 12.8% of the variance. And the first three components explained about 25% of the variance. :(